bert/src/mosaic_bert.py

```diff
@@ -0,0 +1,246 @@
+
+    This function creates a Mosaic BERT, which includes several throughput optimizations not available in |:hugging_face:| BERT as well
+        as architecture changes based on ALiBi and Gated Linear Units.
+
+    Args:
+        pretrained_model_name (str): Name of the Hugging Face model to instantiate. This will determine the default model configuration.
+            Default: ``bert-base-uncased``.
+        model_config (dict): A dictionary of user-specified configurations to update/add to the default model configuration.
+        tokenizer_name (str, optional): Tokenizer name used to preprocess the dataset and validate the models inputs.
+        gradient_checkpointing (bool, optional): Use gradient checkpointing. Default: ``False``.
+        pretrained_checkpoint (str, optional): The pretrained checkpoint to initialize the model weights. If provided, 
+            the state dictionary stored at `pretrained_checkpoint` will be loaded into the model after initialization. Default: ``None``.
+
+    .. code-block::
+
+            {
+              "_name_or_path": "bert-base-uncased",
+              "architectures": ["BertForMaskedLM"],
+              "attention_probs_dropout_prob": 0.0,
+              "classifier_dropout": null,
+              "gradient_checkpointing": false,
+              "hidden_act": "gelu",
+              "hidden_dropout_prob": 0.1,
+              "hidden_size": 768,
+              "initializer_range": 0.02,
+              "intermediate_size": 3072,
+              "layer_norm_eps": 1e-12,
+              "max_position_embeddings": 512,
+              "model_type": "bert",
+              "num_attention_heads": 12,
+              "num_hidden_layers": 12,
+              "pad_token_id": 0,
+              "position_embedding_type": "absolute",
+              "transformers_version": "4.16.0",
+              "type_vocab_size": 2,
+              "use_cache": true,
+              "vocab_size": 30522
+            }
+
+   To create a Mosaic BERT model for Masked Language Model pretraining:
+
+    .. testcode::
+
+        from src.mosaic import create_mosaic_bert_mlm
+        model = create_mosaic_bert_mlm()
+
+    """
+    if not model_config:
+        model_config = {}
+
+    # By default, turn off attention dropout in Mosaic BERT (otherwise, Flash Attention will be off by default)
+    if 'attention_probs_dropout_prob' not in model_config:
+        model_config['attention_probs_dropout_prob'] = 0.0
+
+    if not pretrained_model_name:
+        pretrained_model_name = 'bert-base-uncased'
+
+    config = transformers.AutoConfig.from_pretrained(pretrained_model_name, **model_config)
+    config.return_dict = False
+    # Padding for divisibility by 8
+    if config.vocab_size % 8 != 0:
+        config.vocab_size += 8 - (config.vocab_size % 8)
+
+    if pretrained_checkpoint is not None:
+        model = BertForMaskedLM.from_pretrained(pretrained_checkpoint=pretrained_checkpoint, config=config)
```

Comment:
Is the `pretrained_checkpoint` here a composer checkpoint? If so, it won't work because the keys will have the extra model prefix right? I could definitely be missing something here.

Reply:
No, not a composer checkpoint. This kind of HF API loads in a saved torch state_dict, so `pretrained_checkpoint` is the path to that save file. The format of the state_dict (i.e. the prefixes) would depend on how the user created the save file. We could anticipate the likely use case that they would save from the wrapped `ComposerModel`, meaning it would have the extra `model` prefix. So maybe the `from_pretrained` method should remove any such leading prefix.